File: app/src/main/cpp/wn-steam-client/src/vdf.cpp
#include "wn_steam/vdf.h"

#include <algorithm>
#include <cstring>
#include <cctype>
#include <charconv>

namespace wn_steam::vdf {

namespace {

constexpr uint8_t kTypeNone        = 0x00;
constexpr uint8_t kTypeString      = 0x01;
constexpr uint8_t kTypeInt32       = 0x02;
constexpr uint8_t kTypeFloat32     = 0x03;
constexpr uint8_t kTypePointer     = 0x04;
constexpr uint8_t kTypeWideString  = 0x05;
constexpr uint8_t kTypeColor       = 0x06;
constexpr uint8_t kTypeUInt64      = 0x07;
constexpr uint8_t kTypeEnd         = 0x08;
constexpr uint8_t kTypeInt64       = 0x09;
constexpr uint8_t kTypeEndAlt      = 0x0B;

[[nodiscard]] bool iequals(std::string_view a, std::string_view b) noexcept {
    if (a.size() != b.size()) return false;
    for (size_t i = 0; i < a.size(); ++i) {
        if (std::tolower(static_cast<unsigned char>(a[i])) !=
            std::tolower(static_cast<unsigned char>(b[i]))) {
            return false;
        }
    }
    return true;
}

// Walking cursor over the binary stream. Bounds-checked: any out-of-range
// read sets `ok_ = false` and returns a zero/empty value so the caller can
// drop the partial result.
class Cursor {
public:
    explicit Cursor(std::span<const uint8_t> buf) noexcept : buf_(buf) {}

    [[nodiscard]] bool ok()  const noexcept { return ok_; }
    [[nodiscard]] bool eof() const noexcept { return pos_ >= buf_.size(); }

    [[nodiscard]] uint8_t read_u8() noexcept {
        if (pos_ + 1 > buf_.size()) { ok_ = false; return 0; }
        return buf_[pos_++];
    }
    [[nodiscard]] uint32_t read_u32_le() noexcept {
        if (pos_ + 4 > buf_.size()) { ok_ = false; return 0; }
        uint32_t v = static_cast<uint32_t>(buf_[pos_])
                   | (static_cast<uint32_t>(buf_[pos_ + 1]) << 8)
                   | (static_cast<uint32_t>(buf_[pos_ + 2]) << 16)
                   | (static_cast<uint32_t>(buf_[pos_ + 3]) << 24);
        pos_ += 4;
        return v;
    }
    [[nodiscard]] uint64_t read_u64_le() noexcept {
        uint64_t lo = read_u32_le();
        uint64_t hi = read_u32_le();
        return lo | (hi << 32);
    }
    [[nodiscard]] float read_f32_le() noexcept {
        uint32_t bits = read_u32_le();
        float out;
        std::memcpy(&out, &bits, sizeof(out));
        return out;
    }
    [[nodiscard]] std::string read_cstring() noexcept {
        std::string out;
        while (pos_ < buf_.size()) {
            uint8_t b = buf_[pos_++];
            if (b == 0) return out;
            out.push_back(static_cast<char>(b));
        }
        ok_ = false;
        return out;
    }
    [[nodiscard]] std::u16string read_wide_cstring() noexcept {
        std::u16string out;
        while (pos_ + 1 < buf_.size()) {
            uint16_t u = static_cast<uint16_t>(buf_[pos_])
                       | (static_cast<uint16_t>(buf_[pos_ + 1]) << 8);
            pos_ += 2;
            if (u == 0) return out;
            out.push_back(static_cast<char16_t>(u));
        }
        ok_ = false;
        return out;
    }
    void skip(size_t n) noexcept {
        if (pos_ + n > buf_.size()) { ok_ = false; return; }
        pos_ += n;
    }

private:
    std::span<const uint8_t> buf_;
    size_t                   pos_ = 0;
    bool                     ok_  = true;
};

// Parse a single record's value (type-tag already consumed). The key has
// also already been read. On a None tag the caller passes a freshly created
// node so children get appended directly.
[[nodiscard]] bool parse_value(Cursor& c, uint8_t type, KVNode& node) {
    switch (type) {
        case kTypeNone: {
            // Subobject — recurse, appending children until we hit End.
            while (c.ok()) {
                uint8_t inner_type = c.read_u8();
                if (!c.ok()) return false;
                if (inner_type == kTypeEnd || inner_type == kTypeEndAlt) return true;
                auto child = std::make_unique<KVNode>(c.read_cstring());
                if (!c.ok()) return false;
                if (!parse_value(c, inner_type, *child)) return false;
                node.add_child(std::move(child));
            }
            return false;
        }
        case kTypeString:
            node.set_value(c.read_cstring());
            return c.ok();
        case kTypeInt32:
            node.set_value(static_cast<int32_t>(c.read_u32_le()));
            return c.ok();
        case kTypeFloat32:
            node.set_value(c.read_f32_le());
            return c.ok();
        case kTypePointer:
        case kTypeColor:
            node.set_value(c.read_u32_le());
            return c.ok();
        case kTypeWideString:
            node.set_value(c.read_wide_cstring());
            return c.ok();
        case kTypeUInt64:
            node.set_value(c.read_u64_le());
            return c.ok();
        case kTypeInt64:
            node.set_value(static_cast<int64_t>(c.read_u64_le()));
            return c.ok();
        default:
            // Unknown type — corrupt or version we don't yet understand.
            return false;
    }
}

}  // namespace

const KVNode* KVNode::child(std::string_view key) const noexcept {
    for (const auto& c : children_) {
        if (iequals(c->name(), key)) return c.get();
    }
    return nullptr;
}

std::string KVNode::as_string(std::string_view fallback) const {
    return std::visit(
        [&fallback](auto&& v) -> std::string {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, std::string>)        return v;
            else if constexpr (std::is_same_v<T, int32_t>)       return std::to_string(v);
            else if constexpr (std::is_same_v<T, int64_t>)       return std::to_string(v);
            else if constexpr (std::is_same_v<T, uint32_t>)      return std::to_string(v);
            else if constexpr (std::is_same_v<T, uint64_t>)      return std::to_string(v);
            else if constexpr (std::is_same_v<T, float>)         return std::to_string(v);
            else                                                  return std::string(fallback);
        },
        value_);
}

int64_t KVNode::as_int(int64_t fallback) const noexcept {
    return std::visit(
        [fallback](auto&& v) -> int64_t {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, int32_t>)  return v;
            else if constexpr (std::is_same_v<T, int64_t>)  return v;
            else if constexpr (std::is_same_v<T, uint32_t>) return static_cast<int64_t>(v);
            else if constexpr (std::is_same_v<T, uint64_t>) return static_cast<int64_t>(v);
            else if constexpr (std::is_same_v<T, std::string>) {
                int64_t out = 0;
                auto* first = v.data();
                auto* last  = v.data() + v.size();
                auto r = std::from_chars(first, last, out);
                return (r.ec == std::errc{}) ? out : fallback;
            }
            else return fallback;
        },
        value_);
}

uint64_t KVNode::as_uint(uint64_t fallback) const noexcept {
    return std::visit(
        [fallback](auto&& v) -> uint64_t {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, uint32_t>) return v;
            else if constexpr (std::is_same_v<T, uint64_t>) return v;
            else if constexpr (std::is_same_v<T, int32_t>)  return static_cast<uint64_t>(v);
            else if constexpr (std::is_same_v<T, int64_t>)  return static_cast<uint64_t>(v);
            else if constexpr (std::is_same_v<T, std::string>) {
                uint64_t out = 0;
                auto* first = v.data();
                auto* last  = v.data() + v.size();
                auto r = std::from_chars(first, last, out);
                return (r.ec == std::errc{}) ? out : fallback;
            }
            else return fallback;
        },
        value_);
}

bool KVNode::as_bool(bool fallback) const noexcept {
    return std::visit(
        [fallback](auto&& v) -> bool {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, int32_t>  || std::is_same_v<T, int64_t> ||
                          std::is_same_v<T, uint32_t> || std::is_same_v<T, uint64_t>) {
                return v != 0;
            } else if constexpr (std::is_same_v<T, std::string>) {
                if (v == "1" || iequals(v, "true") || iequals(v, "yes")) return true;
                if (v == "0" || iequals(v, "false") || iequals(v, "no")) return false;
                return fallback;
            } else {
                return fallback;
            }
        },
        value_);
}

// =====================================================================
// Text VDF
// =====================================================================
namespace {

struct TextCursor {
    std::span<const uint8_t> buf;
    size_t pos = 0;
    bool   ok  = true;

    [[nodiscard]] bool eof() const noexcept { return pos >= buf.size(); }
    [[nodiscard]] char peek() const noexcept {
        return pos < buf.size() ? static_cast<char>(buf[pos]) : '\0';
    }
    char next() noexcept {
        return pos < buf.size() ? static_cast<char>(buf[pos++]) : '\0';
    }
    void advance() noexcept { if (pos < buf.size()) ++pos; }

    void skip_ws_and_comments() noexcept {
        while (pos < buf.size()) {
            char c = static_cast<char>(buf[pos]);
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r') { ++pos; continue; }
            // Line comment: // ... \n
            if (c == '/' && pos + 1 < buf.size() && buf[pos + 1] == '/') {
                pos += 2;
                while (pos < buf.size() && buf[pos] != '\n') ++pos;
                continue;
            }
            // Block comment: /* ... */
            if (c == '/' && pos + 1 < buf.size() && buf[pos + 1] == '*') {
                pos += 2;
                while (pos + 1 < buf.size() &&
                       !(buf[pos] == '*' && buf[pos + 1] == '/')) {
                    ++pos;
                }
                pos = std::min(buf.size(), pos + 2);
                continue;
            }
            break;
        }
    }

    // Read one token: either a quoted string (supporting common backslash
    // escapes \n \t \r \\ \") or an unquoted identifier (continues until
    // whitespace, brace, or quote). Returns nullopt at EOF.
    [[nodiscard]] std::optional<std::string> read_token() noexcept {
        skip_ws_and_comments();
        if (eof()) return std::nullopt;
        if (peek() == '"') {
            advance();
            std::string out;
            while (!eof()) {
                char c = next();
                if (c == '"') return out;
                if (c == '\\' && !eof()) {
                    char esc = next();
                    switch (esc) {
                        case 'n':  out.push_back('\n'); break;
                        case 't':  out.push_back('\t'); break;
                        case 'r':  out.push_back('\r'); break;
                        case '\\': out.push_back('\\'); break;
                        case '"':  out.push_back('"');  break;
                        default:   out.push_back(esc);  break;
                    }
                } else {
                    out.push_back(c);
                }
            }
            ok = false;  // unterminated string
            return std::nullopt;
        }
        std::string out;
        while (!eof()) {
            char c = peek();
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r' ||
                c == '{' || c == '}' || c == '"') break;
            out.push_back(next());
        }
        if (out.empty()) return std::nullopt;
        return out;
    }
};

// Parse the body of an object until `}` or EOF.
[[nodiscard]] bool parse_text_object(TextCursor& c, KVNode& parent) noexcept {
    while (true) {
        c.skip_ws_and_comments();
        if (c.eof()) return true;
        if (c.peek() == '}') { c.advance(); return true; }
        auto key = c.read_token();
        if (!key) return c.ok;
        c.skip_ws_and_comments();
        if (c.eof()) return false;
        auto child = std::make_unique<KVNode>(std::move(*key));
        if (c.peek() == '{') {
            c.advance();
            if (!parse_text_object(c, *child)) return false;
        } else {
            auto val = c.read_token();
            if (!val) return false;
            child->set_value(std::move(*val));
        }
        parent.add_child(std::move(child));
    }
}

}  // namespace

KVNodePtr parse_text(std::span<const uint8_t> body) noexcept {
    TextCursor c{body, 0, true};
    c.skip_ws_and_comments();
    if (c.eof()) return nullptr;
    auto key = c.read_token();
    if (!key) return nullptr;
    c.skip_ws_and_comments();
    if (c.peek() != '{') return nullptr;
    c.advance();
    auto root = std::make_unique<KVNode>(std::move(*key));
    if (!parse_text_object(c, *root)) return nullptr;
    return root;
}

KVNodePtr parse_auto(std::span<const uint8_t> body) noexcept {
    if (body.empty()) return nullptr;
    size_t i = 0;
    while (i < body.size() &&
           (body[i] == ' ' || body[i] == '\t' ||
            body[i] == '\n' || body[i] == '\r')) ++i;
    if (i >= body.size()) return nullptr;
    if (body[i] == '"') return parse_text(body);
    return parse_binary(body);
}

KVNodePtr parse_binary(std::span<const uint8_t> body) noexcept {
    if (body.empty()) return nullptr;
    Cursor c(body);
    uint8_t type = c.read_u8();
    if (!c.ok()) return nullptr;
    auto root = std::make_unique<KVNode>(c.read_cstring());
    if (!c.ok()) return nullptr;
    if (!parse_value(c, type, *root)) return nullptr;
    return root;
}

KVNodePtr parse_binary_package(std::span<const uint8_t> body,
                               uint32_t* out_packageid_prefix) noexcept {
    if (body.size() < 4) return nullptr;
    uint32_t prefix = static_cast<uint32_t>(body[0])
                    | (static_cast<uint32_t>(body[1]) << 8)
                    | (static_cast<uint32_t>(body[2]) << 16)
                    | (static_cast<uint32_t>(body[3]) << 24);
    if (out_packageid_prefix) *out_packageid_prefix = prefix;
    return parse_binary(body.subspan(4));
}

}  // namespace wn_steam::vdf

File: app/src/main/cpp/wn-steam-client/rust/src/vdf.rs
#[derive(Clone, Debug, PartialEq)]
pub enum KVValue {
    Object,
    String(String),
    Int32(i32),
    Float32(f32),
    UInt32(u32),
    WideString(Vec<u16>),
    UInt64(u64),
    Int64(i64),
}

#[derive(Clone, Debug, PartialEq)]
pub struct KVNode {
    pub name: String,
    pub value: KVValue,
    pub children: Vec<KVNode>,
}

impl KVNode {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            value: KVValue::Object,
            children: Vec::new(),
        }
    }

    pub fn is_object(&self) -> bool {
        matches!(self.value, KVValue::Object)
    }

    pub fn child(&self, key: &str) -> Option<&KVNode> {
        self.children
            .iter()
            .find(|c| c.name.eq_ignore_ascii_case(key))
    }

    pub fn as_string(&self, fallback: &str) -> String {
        match &self.value {
            KVValue::String(v) => v.clone(),
            KVValue::Int32(v) => v.to_string(),
            KVValue::Float32(v) => v.to_string(),
            KVValue::UInt32(v) => v.to_string(),
            KVValue::UInt64(v) => v.to_string(),
            KVValue::Int64(v) => v.to_string(),
            _ => fallback.to_string(),
        }
    }

    pub fn as_int(&self, fallback: i64) -> i64 {
        match &self.value {
            KVValue::Int32(v) => *v as i64,
            KVValue::Int64(v) => *v,
            KVValue::UInt32(v) => *v as i64,
            KVValue::UInt64(v) => *v as i64,
            KVValue::String(v) => v.parse().unwrap_or(fallback),
            _ => fallback,
        }
    }

    pub fn as_uint(&self, fallback: u64) -> u64 {
        match &self.value {
            KVValue::UInt32(v) => *v as u64,
            KVValue::UInt64(v) => *v,
            KVValue::Int32(v) => *v as u64,
            KVValue::Int64(v) => *v as u64,
            KVValue::String(v) => v.parse().unwrap_or(fallback),
            _ => fallback,
        }
    }

    pub fn as_bool(&self, fallback: bool) -> bool {
        match &self.value {
            KVValue::Int32(v) => *v != 0,
            KVValue::Int64(v) => *v != 0,
            KVValue::UInt32(v) => *v != 0,
            KVValue::UInt64(v) => *v != 0,
            KVValue::String(v)
                if v == "1" || v.eq_ignore_ascii_case("true") || v.eq_ignore_ascii_case("yes") =>
            {
                true
            }
            KVValue::String(v)
                if v == "0" || v.eq_ignore_ascii_case("false") || v.eq_ignore_ascii_case("no") =>
            {
                false
            }
            _ => fallback,
        }
    }
}

const TYPE_NONE: u8 = 0x00;
const TYPE_STRING: u8 = 0x01;
const TYPE_INT32: u8 = 0x02;
const TYPE_FLOAT32: u8 = 0x03;
const TYPE_POINTER: u8 = 0x04;
const TYPE_WIDE_STRING: u8 = 0x05;
const TYPE_COLOR: u8 = 0x06;
const TYPE_UINT64: u8 = 0x07;
const TYPE_END: u8 = 0x08;
const TYPE_INT64: u8 = 0x09;
const TYPE_END_ALT: u8 = 0x0b;

struct Cursor<'a> {
    buf: &'a [u8],
    pos: usize,
    ok: bool,
}

impl<'a> Cursor<'a> {
    fn new(buf: &'a [u8]) -> Self {
        Self {
            buf,
            pos: 0,
            ok: true,
        }
    }

    fn read_u8(&mut self) -> u8 {
        if self.pos + 1 > self.buf.len() {
            self.ok = false;
            return 0;
        }
        let v = self.buf[self.pos];
        self.pos += 1;
        v
    }

    fn read_u32_le(&mut self) -> u32 {
        if self.pos + 4 > self.buf.len() {
            self.ok = false;
            return 0;
        }
        let v = u32::from_le_bytes(self.buf[self.pos..self.pos + 4].try_into().unwrap());
        self.pos += 4;
        v
    }

    fn read_u64_le(&mut self) -> u64 {
        let lo = self.read_u32_le() as u64;
        let hi = self.read_u32_le() as u64;
        lo | (hi << 32)
    }

    fn read_f32_le(&mut self) -> f32 {
        f32::from_bits(self.read_u32_le())
    }

    fn read_cstring(&mut self) -> String {
        let mut out = Vec::new();
        while self.pos < self.buf.len() {
            let b = self.buf[self.pos];
            self.pos += 1;
            if b == 0 {
                return String::from_utf8_lossy(&out).into_owned();
            }
            out.push(b);
        }
        self.ok = false;
        String::from_utf8_lossy(&out).into_owned()
    }

    fn read_wide_cstring(&mut self) -> Vec<u16> {
        let mut out = Vec::new();
        while self.pos + 1 < self.buf.len() {
            let u = u16::from_le_bytes([self.buf[self.pos], self.buf[self.pos + 1]]);
            self.pos += 2;
            if u == 0 {
                return out;
            }
            out.push(u);
        }
        self.ok = false;
        out
    }
}

fn parse_value(cursor: &mut Cursor<'_>, ty: u8, node: &mut KVNode) -> bool {
    match ty {
        TYPE_NONE => {
            while cursor.ok {
                let inner_type = cursor.read_u8();
                if !cursor.ok {
                    return false;
                }
                if inner_type == TYPE_END || inner_type == TYPE_END_ALT {
                    return true;
                }
                let name = cursor.read_cstring();
                if !cursor.ok {
                    return false;
                }
                let mut child = KVNode::new(name);
                if !parse_value(cursor, inner_type, &mut child) {
                    return false;
                }
                node.children.push(child);
            }
            false
        }
        TYPE_STRING => {
            node.value = KVValue::String(cursor.read_cstring());
            cursor.ok
        }
        TYPE_INT32 => {
            node.value = KVValue::Int32(cursor.read_u32_le() as i32);
            cursor.ok
        }
        TYPE_FLOAT32 => {
            node.value = KVValue::Float32(cursor.read_f32_le());
            cursor.ok
        }
        TYPE_POINTER | TYPE_COLOR => {
            node.value = KVValue::UInt32(cursor.read_u32_le());
            cursor.ok
        }
        TYPE_WIDE_STRING => {
            node.value = KVValue::WideString(cursor.read_wide_cstring());
            cursor.ok
        }
        TYPE_UINT64 => {
            node.value = KVValue::UInt64(cursor.read_u64_le());
            cursor.ok
        }
        TYPE_INT64 => {
            node.value = KVValue::Int64(cursor.read_u64_le() as i64);
            cursor.ok
        }
        _ => false,
    }
}

pub fn parse_binary(body: &[u8]) -> Option<KVNode> {
    if body.is_empty() {
        return None;
    }
    let mut cursor = Cursor::new(body);
    let ty = cursor.read_u8();
    if !cursor.ok {
        return None;
    }
    let mut root = KVNode::new(cursor.read_cstring());
    if !cursor.ok || !parse_value(&mut cursor, ty, &mut root) {
        return None;
    }
    Some(root)
}

pub fn parse_binary_package(body: &[u8]) -> Option<(u32, KVNode)> {
    if body.len() < 4 {
        return None;
    }
    let package_id = u32::from_le_bytes(body[..4].try_into().unwrap());
    Some((package_id, parse_binary(&body[4..])?))
}

pub fn parse_auto(body: &[u8]) -> Option<KVNode> {
    let first = body
        .iter()
        .copied()
        .find(|b| !matches!(b, b' ' | b'\t' | b'\n' | b'\r'))?;
    if first == b'"' {
        parse_text(body)
    } else {
        parse_binary(body)
    }
}

struct TextCursor<'a> {
    buf: &'a [u8],
    pos: usize,
    ok: bool,
}

impl<'a> TextCursor<'a> {
    fn new(buf: &'a [u8]) -> Self {
        Self {
            buf,
            pos: 0,
            ok: true,
        }
    }

    fn eof(&self) -> bool {
        self.pos >= self.buf.len()
    }

    fn peek(&self) -> u8 {
        self.buf.get(self.pos).copied().unwrap_or(0)
    }

    fn next(&mut self) -> u8 {
        let b = self.peek();
        if !self.eof() {
            self.pos += 1;
        }
        b
    }

    fn skip_ws_and_comments(&mut self) {
        while self.pos < self.buf.len() {
            match self.buf[self.pos] {
                b' ' | b'\t' | b'\n' | b'\r' => self.pos += 1,
                b'/' if self.buf.get(self.pos + 1) == Some(&b'/') => {
                    self.pos += 2;
                    while self.pos < self.buf.len() && self.buf[self.pos] != b'\n' {
                        self.pos += 1;
                    }
                }
                b'/' if self.buf.get(self.pos + 1) == Some(&b'*') => {
                    self.pos += 2;
                    while self.pos + 1 < self.buf.len()
                        && !(self.buf[self.pos] == b'*' && self.buf[self.pos + 1] == b'/')
                    {
                        self.pos += 1;
                    }
                    self.pos = (self.pos + 2).min(self.buf.len());
                }
                _ => break,
            }
        }
    }

    fn read_token(&mut self) -> Option<String> {
        self.skip_ws_and_comments();
        if self.eof() {
            return None;
        }
        if self.peek() == b'"' {
            self.pos += 1;
            let mut out = Vec::new();
            while !self.eof() {
                let c = self.next();
                if c == b'"' {
                    return Some(String::from_utf8_lossy(&out).into_owned());
                }
                if c == b'\\' && !self.eof() {
                    out.push(match self.next() {
                        b'n' => b'\n',
                        b't' => b'\t',
                        b'r' => b'\r',
                        b'\\' => b'\\',
                        b'"' => b'"',
                        other => other,
                    });
                } else {
                    out.push(c);
                }
            }
            self.ok = false;
            return None;
        }

        let start = self.pos;
        while !self.eof() {
            let c = self.peek();
            if matches!(c, b' ' | b'\t' | b'\n' | b'\r' | b'{' | b'}' | b'"') {
                break;
            }
            self.pos += 1;
        }
        (self.pos > start).then(|| String::from_utf8_lossy(&self.buf[start..self.pos]).into_owned())
    }
}

fn parse_text_object(cursor: &mut TextCursor<'_>, parent: &mut KVNode) -> bool {
    loop {
        cursor.skip_ws_and_comments();
        if cursor.eof() {
            return true;
        }
        if cursor.peek() == b'}' {
            cursor.pos += 1;
            return true;
        }
        let Some(key) = cursor.read_token() else {
            return cursor.ok;
        };
        cursor.skip_ws_and_comments();
        if cursor.eof() {
            return false;
        }
        let mut child = KVNode::new(key);
        if cursor.peek() == b'{' {
            cursor.pos += 1;
            if !parse_text_object(cursor, &mut child) {
                return false;
            }
        } else {
            let Some(value) = cursor.read_token() else {
                return false;
            };
            child.value = KVValue::String(value);
        }
        parent.children.push(child);
    }
}

pub fn parse_text(body: &[u8]) -> Option<KVNode> {
    let mut cursor = TextCursor::new(body);
    cursor.skip_ws_and_comments();
    let key = cursor.read_token()?;
    cursor.skip_ws_and_comments();
    if cursor.peek() != b'{' {
        return None;
    }
    cursor.pos += 1;
    let mut root = KVNode::new(key);
    parse_text_object(&mut cursor, &mut root).then_some(root)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_text_vdf_with_comments_and_case_insensitive_lookup() {
        let text = br#"
            // comment
            "appinfo" {
                "appid" "480"
                "Common" { "name" "Spacewar" "type" "Game" }
                /* block */ "enabled" "true"
            }
        "#;
        let root = parse_text(text).unwrap();
        assert_eq!(root.name, "appinfo");
        assert_eq!(root.child("APPID").unwrap().as_int(0), 480);
        assert_eq!(
            root.child("common")
                .unwrap()
                .child("NAME")
                .unwrap()
                .as_string(""),
            "Spacewar"
        );
        assert!(root.child("enabled").unwrap().as_bool(false));
    }

    #[test]
    fn parses_binary_tree() {
        let mut bytes = Vec::new();
        bytes.extend_from_slice(&[TYPE_NONE]);
        bytes.extend_from_slice(b"root\0");
        bytes.extend_from_slice(&[TYPE_STRING]);
        bytes.extend_from_slice(b"name\0Spacewar\0");
        bytes.extend_from_slice(&[TYPE_INT32]);
        bytes.extend_from_slice(b"appid\0");
        bytes.extend_from_slice(&480i32.to_le_bytes());
        bytes.extend_from_slice(&[TYPE_END]);

        let root = parse_binary(&bytes).unwrap();
        assert!(root.is_object());
        assert_eq!(root.child("name").unwrap().as_string(""), "Spacewar");
        assert_eq!(root.child("appid").unwrap().as_int(0), 480);
    }

    #[test]
    fn parses_package_prefix() {
        let mut bytes = 123u32.to_le_bytes().to_vec();
        bytes.extend_from_slice(&[TYPE_NONE]);
        bytes.extend_from_slice(b"package\0");
        bytes.extend_from_slice(&[TYPE_END]);
        let (package_id, root) = parse_binary_package(&bytes).unwrap();
        assert_eq!(package_id, 123);
        assert_eq!(root.name, "package");
    }

    // Manifest gid/size/download: proves the parser reads download (the 3rd value)
    // correctly, ruling out a parser cause for the stale corrupt download sizes.
    #[test]
    fn parses_manifest_uint64_gid_size_download() {
        let gid: u64 = 8072044898226043193;
        let size: u64 = 30738676601;
        let download: u64 = 15000000000;
        let mut b = vec![TYPE_NONE];
        b.extend_from_slice(b"public\0");
        b.push(TYPE_UINT64);
        b.extend_from_slice(b"gid\0");
        b.extend_from_slice(&gid.to_le_bytes());
        b.push(TYPE_UINT64);
        b.extend_from_slice(b"size\0");
        b.extend_from_slice(&size.to_le_bytes());
        b.push(TYPE_UINT64);
        b.extend_from_slice(b"download\0");
        b.extend_from_slice(&download.to_le_bytes());
        b.push(TYPE_END);

        let root = parse_binary(&b).unwrap();
        assert_eq!(root.child("gid").unwrap().as_string(""), gid.to_string());
        assert_eq!(root.child("size").unwrap().as_string(""), size.to_string());
        assert_eq!(
            root.child("download").unwrap().as_string(""),
            download.to_string()
        );
    }

    #[test]
    fn parses_manifest_text_vdf() {
        let text = br#"
            "public"
            {
                "gid"		"8072044898226043193"
                "size"		"30738676601"
                "download"		"15000000000"
            }
        "#;
        let root = parse_text(text).unwrap();
        assert_eq!(root.child("size").unwrap().as_string(""), "30738676601");
        assert_eq!(root.child("download").unwrap().as_string(""), "15000000000");
    }
}
